File: SimulManager/SimulManager/Services/Clock/VirtualClock.cs
// 가상시계·가속엔진 — 시퀀스 설계서 §1.3·§3 (자유진행 원칙)
// (backend/app/clock/virtual_clock.py 포팅).
//
// - Run 시작(P3)~종료(P5)까지 노드 접속 여부와 무관하게 연속 진행 (free-running)
// - 가상 1분 = 실 (가속배율/15) 초. 예: 배율 10 s → 실 0.667 s
// - '일시정지'는 운영자의 명시적 조작에 한해서만 사용 (자동 예외처리 금지)
// - 30일차 마지막 구간(인덱스 95) 완료 시점에 정지 (P5) → COMPLETED
// - P6 케이스 전환 시 Rewind()로 시작시각 되감기
//
// 틱 이벤트 순서(구간 경계 원자성의 기반): 1분 진행 → clock.minute →
// (구간 변화 시) clock.interval → (자정 시) clock.day → (완료 시) clock.complete.
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;

namespace SimulManager.Services.Clock
{
    public enum ClockState
    {
        Idle,       // 시작 전 / 리셋(되감기) 후
        Running,
        Paused,     // 운영자 일시정지
        Completed,  // 전 구간 완료 (P5)
        Stopped,    // 운영자 수동 정지
    }

    public sealed class ClockTickPayload
    {
        public DateTime VirtualTime;
        public int IntervalIndex;
        public int DayNumber;
        public int ElapsedMinutes;
    }

    public sealed class ClockStatus
    {
        public ClockState State;
        public DateTime VirtualTime;
        public int IntervalIndex;
        public int DayNumber;
        public int TotalDays;
        public int ElapsedMinutes;
        public int TotalMinutes;
        public double Progress;
        public double AccelSecondsPer15Min;
        public double TickPeriodS;

        public string StateName => State.ToString().ToLowerInvariant();
    }

    public sealed class VirtualClock
    {
        public const int MinutesPerInterval = 15;
        public const int IntervalsPerDay = 96;

        private SimulationConfig _sim;
        private readonly EventBus _bus;
        private readonly object _lock = new object();
        private ClockState _state = ClockState.Idle;
        private DateTime _virtualTime;
        private int _elapsedMinutes;
        private Thread _thread;
        private readonly ManualResetEventSlim _resume = new ManualResetEventSlim(true);
        private volatile bool _loopCancel;

        public VirtualClock(SimulationConfig sim, EventBus bus)
        {
            _sim = sim;
            _bus = bus;
            _virtualTime = sim.StartVirtualTime;
        }

        // ── 조회 ──────────────────────────────────────────────
        public ClockState State { get { lock (_lock) return _state; } }
        public DateTime VirtualTime { get { lock (_lock) return _virtualTime; } }

        /// <summary>하루 내 15분 구간 인덱스 0~95.</summary>
        public int IntervalIndex
        {
            get { lock (_lock) return IntervalOf(_virtualTime); }
        }

        private static int IntervalOf(DateTime t) => (t.Hour * 60 + t.Minute) / MinutesPerInterval;

        /// <summary>Run 내 일차 (1부터).</summary>
        public int DayNumber
        {
            get { lock (_lock) return (int)(_virtualTime.Date - _sim.StartVirtualTime.Date).TotalDays + 1; }
        }

        public int TotalMinutes => _sim.Days * 24 * 60;
        public double TickPeriodS => _sim.RealSecondsPerVirtualMinute;

        public ClockStatus Status()
        {
            lock (_lock)
            {
                return new ClockStatus
                {
                    State = _state,
                    VirtualTime = _virtualTime,
                    IntervalIndex = IntervalOf(_virtualTime),
                    DayNumber = (int)(_virtualTime.Date - _sim.StartVirtualTime.Date).TotalDays + 1,
                    TotalDays = _sim.Days,
                    ElapsedMinutes = _elapsedMinutes,
                    TotalMinutes = TotalMinutes,
                    Progress = Math.Round((double)_elapsedMinutes / TotalMinutes, 4),
                    AccelSecondsPer15Min = _sim.AccelSecondsPer15Min,
                    TickPeriodS = Math.Round(TickPeriodS, 4),
                };
            }
        }

        // ── 제어 (UI → P3/P5/P6) ─────────────────────────────
        /// <summary>P3 운전 개시 — IDLE에서만 시작.</summary>
        public void Start()
        {
            lock (_lock)
            {
                if (_state != ClockState.Idle)
                    throw new InvalidOperationException("IDLE 상태에서만 시작 가능 (현재: " + _state + ")");
                _state = ClockState.Running;
            }
            _resume.Set();
            _loopCancel = false;
            _thread = new Thread(RunLoop) { IsBackground = true, Name = "virtual-clock" };
            _thread.Start();
            PublishState();
            Trace.TraceInformation("가상시계 시작: " + VirtualTime + ", 배율 " + _sim.AccelSecondsPer15Min + "s/15분");
        }

        /// <summary>운영자 수동 정지.</summary>
        public void Stop()
        {
            lock (_lock)
            {
                if (_state != ClockState.Running && _state != ClockState.Paused)
                    throw new InvalidOperationException("RUNNING/PAUSED 상태에서만 정지 가능 (현재: " + _state + ")");
                _state = ClockState.Stopped;
            }
            _resume.Set();
            CancelLoop();
            PublishState();
        }

        /// <summary>운영자 일시정지 — 자동 예외처리로 호출 금지 (자유진행 원칙).</summary>
        public void Pause()
        {
            lock (_lock)
            {
                if (_state != ClockState.Running)
                    throw new InvalidOperationException("RUNNING 상태에서만 일시정지 가능 (현재: " + _state + ")");
                _state = ClockState.Paused;
            }
            _resume.Reset();
            PublishState();
        }

        public void Resume()
        {
            lock (_lock)
            {
                if (_state != ClockState.Paused)
                    throw new InvalidOperationException("PAUSED 상태에서만 재개 가능 (현재: " + _state + ")");
                _state = ClockState.Running;
            }
            _resume.Set();
            PublishState();
        }

        /// <summary>P6 케이스 전환 — 시작시각으로 되감기 (RUNNING 중에는 불가).</summary>
        public void Rewind()
        {
            lock (_lock)
            {
                if (_state == ClockState.Running)
                    throw new InvalidOperationException("RUNNING 중에는 되감기 불가 — 먼저 정지하세요");
            }
            CancelLoop();
            lock (_lock)
            {
                _virtualTime = _sim.StartVirtualTime;
                _elapsedMinutes = 0;
                _state = ClockState.Idle;
            }
            _resume.Set();
            PublishState();
            Trace.TraceInformation("가상시계 되감기: " + VirtualTime);
        }

        /// <summary>설정 대시보드 반영 — IDLE(대기) 상태에서만. 가상시각을 새 시작시각으로 재설정.</summary>
        public void ApplyConfig(SimulationConfig sim)
        {
            lock (_lock)
            {
                if (_state != ClockState.Idle)
                    throw new InvalidOperationException(
                        "설정 변경은 대기(IDLE) 상태에서만 가능 (현재: " + _state + ") — 정지 후 되감기 하세요");
                _sim = sim;
                _virtualTime = sim.StartVirtualTime;
                _elapsedMinutes = 0;
            }
            PublishState();
        }

        public void Shutdown() => CancelLoop();

        // ── 진행 ──────────────────────────────────────────────
        /// <summary>가상 1분 진행 + 이벤트 발행 — 엔진 스레드에서 실행된다.</summary>
        public void AdvanceMinute()
        {
            ClockTickPayload payload;
            bool intervalChanged, dayBoundary, completed;
            lock (_lock)
            {
                int prevInterval = IntervalOf(_virtualTime);
                _virtualTime = _virtualTime.AddMinutes(1);
                _elapsedMinutes += 1;
                payload = new ClockTickPayload
                {
                    VirtualTime = _virtualTime,
                    IntervalIndex = IntervalOf(_virtualTime),
                    DayNumber = (int)(_virtualTime.Date - _sim.StartVirtualTime.Date).TotalDays + 1,
                    ElapsedMinutes = _elapsedMinutes,
                };
                intervalChanged = payload.IntervalIndex != prevInterval;
                dayBoundary = _virtualTime.Hour == 0 && _virtualTime.Minute == 0;
                completed = _elapsedMinutes >= TotalMinutes;
                if (completed) _state = ClockState.Completed;
            }
            _bus.Publish("clock.minute", payload);
            if (intervalChanged) _bus.Publish("clock.interval", payload);
            if (dayBoundary) _bus.Publish("clock.day", payload);
            if (completed)
            {
                _bus.Publish("clock.complete", Status());
                Trace.TraceInformation("Run 완료: 전 " + (_sim.Days * IntervalsPerDay) + "구간 종료");
            }
        }

        /// <summary>드리프트 보정 틱 루프 — 기준 실시각 + n×주기로 스케줄.</summary>
        private void RunLoop()
        {
            double period = TickPeriodS;
            var sw = Stopwatch.StartNew();
            double baseS = 0;
            long n = 0;
            while (!_loopCancel)
            {
                var state = State;
                if (state != ClockState.Running && state != ClockState.Paused) break;
                _resume.Wait();
                if (_loopCancel || State != ClockState.Running) { if (State == ClockState.Paused) continue; break; }
                n++;
                double delay = baseS + n * period - sw.Elapsed.TotalSeconds;
                if (delay > 0)
                    Thread.Sleep(TimeSpan.FromSeconds(delay));
                else if (delay < -period * 10)
                    baseS = sw.Elapsed.TotalSeconds - n * period; // 일시정지 등으로 크게 뒤처짐 — 기준 재설정
                if (_loopCancel || State != ClockState.Running) continue;
                _bus.Invoke(AdvanceMinute); // 엔진 스레드에서 원자적으로 처리 (backlog 방지 대기)
            }
            if (!_loopCancel) PublishState();
        }

        private void CancelLoop()
        {
            _loopCancel = true;
            _resume.Set();
            var t = _thread;
            if (t != null && t.IsAlive && t != Thread.CurrentThread)
                t.Join(2000);
            _thread = null;
        }

        private void PublishState() => _bus.Publish("clock.state", Status());
    }
}

File: backend/app/clock/virtual_clock.py
"""가상시계·가속엔진 — 시퀀스 설계서 §1.3·§3 (자유진행 원칙).

- Run 시작(P3)~종료(P5)까지 노드 접속 여부와 무관하게 연속 진행 (free-running)
- 가상 1분 = 실 (가속배율/15) 초. 예: 배율 10 s → 실 0.667 s
- '일시정지'는 운영자의 명시적 조작에 한해서만 사용 (자동 예외처리 금지)
- 30일차 마지막 구간(인덱스 95) 완료 시점에 정지 (P5) → COMPLETED
- P6 케이스 전환 시 rewind()로 시작시각 되감기

틱 이벤트 순서(구간 경계 원자성의 기반): 1분 진행 → clock.minute →
(구간 변화 시) clock.interval → (자정 시) clock.day → (완료 시) clock.complete.
시나리오 배포자는 clock.minute에서 시각·기상을 먼저 갱신한 뒤 구간 인덱스를 갱신한다.
"""
from __future__ import annotations

import asyncio
import logging
from datetime import datetime, timedelta
from enum import Enum
from typing import Any

from ..config import SimulationConfig
from ..events import EventBus

logger = logging.getLogger(__name__)

MINUTES_PER_INTERVAL = 15
INTERVALS_PER_DAY = 96


class ClockState(str, Enum):
    IDLE = "idle"            # 시작 전 / 리셋(되감기) 후
    RUNNING = "running"
    PAUSED = "paused"        # 운영자 일시정지
    COMPLETED = "completed"  # 전 구간 완료 (P5)
    STOPPED = "stopped"      # 운영자 수동 정지


class VirtualClock:
    def __init__(self, sim: SimulationConfig, bus: EventBus) -> None:
        self._sim = sim
        self._bus = bus
        self._state = ClockState.IDLE
        self._virtual_time = sim.start_virtual_time
        self._elapsed_minutes = 0
        self._task: asyncio.Task | None = None
        self._resume_event = asyncio.Event()
        self._resume_event.set()

    # ── 조회 ──────────────────────────────────────────────
    @property
    def state(self) -> ClockState:
        return self._state

    @property
    def virtual_time(self) -> datetime:
        return self._virtual_time

    @property
    def interval_index(self) -> int:
        """하루 내 15분 구간 인덱스 0~95."""
        return (self._virtual_time.hour * 60 + self._virtual_time.minute) // MINUTES_PER_INTERVAL

    @property
    def day_number(self) -> int:
        """Run 내 일차 (1부터)."""
        return (self._virtual_time - self._sim.start_virtual_time).days + 1

    @property
    def total_minutes(self) -> int:
        return self._sim.days * 24 * 60

    @property
    def tick_period_s(self) -> float:
        return self._sim.real_seconds_per_virtual_minute

    def status(self) -> dict[str, Any]:
        return {
            "state": self._state.value,
            "virtual_time": self._virtual_time.isoformat(),
            "interval_index": self.interval_index,
            "day_number": self.day_number,
            "total_days": self._sim.days,
            "elapsed_minutes": self._elapsed_minutes,
            "total_minutes": self.total_minutes,
            "progress": round(self._elapsed_minutes / self.total_minutes, 4),
            "accel_seconds_per_15min": self._sim.accel_seconds_per_15min,
            "tick_period_s": round(self.tick_period_s, 4),
        }

    # ── 제어 (REST → P3/P5/P6) ────────────────────────────
    async def start(self) -> None:
        """P3 운전 개시 — IDLE에서만 시작."""
        if self._state != ClockState.IDLE:
            raise RuntimeError(f"IDLE 상태에서만 시작 가능 (현재: {self._state.value})")
        self._state = ClockState.RUNNING
        self._resume_event.set()
        self._task = asyncio.create_task(self._run_loop(), name="virtual-clock")
        await self._publish_state()
        logger.info("가상시계 시작: %s, 배율 %ss/15분", self._virtual_time, self._sim.accel_seconds_per_15min)

    async def stop(self) -> None:
        """운영자 수동 정지."""
        if self._state not in (ClockState.RUNNING, ClockState.PAUSED):
            raise RuntimeError(f"RUNNING/PAUSED 상태에서만 정지 가능 (현재: {self._state.value})")
        self._state = ClockState.STOPPED
        self._resume_event.set()
        await self._cancel_task()
        await self._publish_state()
        logger.info("가상시계 수동 정지: %s", self._virtual_time)

    async def pause(self) -> None:
        """운영자 일시정지 — 자동 예외처리로 호출 금지 (자유진행 원칙)."""
        if self._state != ClockState.RUNNING:
            raise RuntimeError(f"RUNNING 상태에서만 일시정지 가능 (현재: {self._state.value})")
        self._state = ClockState.PAUSED
        self._resume_event.clear()
        await self._publish_state()

    async def resume(self) -> None:
        if self._state != ClockState.PAUSED:
            raise RuntimeError(f"PAUSED 상태에서만 재개 가능 (현재: {self._state.value})")
        self._state = ClockState.RUNNING
        self._resume_event.set()
        await self._publish_state()

    async def rewind(self) -> None:
        """P6 케이스 전환 — 시작시각으로 되감기 (RUNNING 중에는 불가)."""
        if self._state == ClockState.RUNNING:
            raise RuntimeError("RUNNING 중에는 되감기 불가 — 먼저 정지하세요")
        await self._cancel_task()
        self._virtual_time = self._sim.start_virtual_time
        self._elapsed_minutes = 0
        self._state = ClockState.IDLE
        self._resume_event.set()
        await self._publish_state()
        logger.info("가상시계 되감기: %s", self._virtual_time)

    async def apply_config(self, sim: SimulationConfig) -> None:
        """설정 대시보드 반영 — IDLE(대기) 상태에서만. 가상시각을 새 시작시각으로 재설정."""
        if self._state != ClockState.IDLE:
            raise RuntimeError(
                f"설정 변경은 대기(IDLE) 상태에서만 가능 (현재: {self._state.value}) — 정지 후 되감기 하세요"
            )
        self._sim = sim
        self._virtual_time = sim.start_virtual_time
        self._elapsed_minutes = 0
        await self._publish_state()
        logger.info("시뮬레이션 설정 반영: 시작 %s, %d일, 배율 %ss/15분",
                    sim.start_virtual_time, sim.days, sim.accel_seconds_per_15min)

    async def shutdown(self) -> None:
        await self._cancel_task()

    # ── 진행 ──────────────────────────────────────────────
    async def advance_minute(self) -> None:
        """가상 1분 진행 + 이벤트 발행. 루프와 테스트가 공용."""
        prev_interval = self.interval_index
        self._virtual_time += timedelta(minutes=1)
        self._elapsed_minutes += 1

        payload = {
            "virtual_time": self._virtual_time.isoformat(),
            "interval_index": self.interval_index,
            "day_number": self.day_number,
            "elapsed_minutes": self._elapsed_minutes,
        }
        await self._bus.publish("clock.minute", payload)
        if self.interval_index != prev_interval:
            await self._bus.publish("clock.interval", payload)
        if self._virtual_time.hour == 0 and self._virtual_time.minute == 0:
            await self._bus.publish("clock.day", payload)

        if self._elapsed_minutes >= self.total_minutes:
            self._state = ClockState.COMPLETED
            await self._bus.publish("clock.complete", self.status())
            logger.info("Run 완료: 전 %d구간 종료", self._sim.days * INTERVALS_PER_DAY)

    async def _run_loop(self) -> None:
        """드리프트 보정 틱 루프 — 기준 실시각 + n×주기로 스케줄."""
        loop = asyncio.get_running_loop()
        period = self.tick_period_s
        base = loop.time()
        n = 0
        try:
            while self._state in (ClockState.RUNNING, ClockState.PAUSED):
                await self._resume_event.wait()
                if self._state != ClockState.RUNNING:
                    break
                n += 1
                delay = base + n * period - loop.time()
                if delay > 0:
                    await asyncio.sleep(delay)
                elif delay < -period * 10:
                    # 일시정지 등으로 크게 뒤처짐 — 기준 재설정 (몰아치기 방지)
                    base = loop.time() - n * period
                if self._state != ClockState.RUNNING:
                    continue
                await self.advance_minute()
            await self._publish_state()
        except asyncio.CancelledError:
            pass

    async def _publish_state(self) -> None:
        await self._bus.publish("clock.state", self.status())

    async def _cancel_task(self) -> None:
        if self._task and not self._task.done():
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        self._task = None
